metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/update/ESBulkProcessor.java

```diff
@@ -109,6 +122,76 @@ public Optional<BulkByScrollResponse> deleteByQuery(QueryBuilder queryBuilder, b
         return Optional.empty();
     }
 
+    public Optional<Cancellable> reindex(QueryBuilder queryBuilder, String index, int batchSize, @Nullable TimeValue timeout) {
+        final long startTime = System.currentTimeMillis();
+        String destIndex = index + "_" + startTime;
+        ReindexRequest reindexRequest = new ReindexRequest()
+            .setSourceQuery(queryBuilder).setDestIndex(destIndex)
+            .setSourceBatchSize(batchSize);
+        if (timeout != null) {
+            reindexRequest.setTimeout(timeout);
+        }
+        reindexRequest.setSourceIndices(index);
+        try {
+            // flush pending writes
+            bulkProcessor.flush();
+            ActionListener<BulkByScrollResponse> renameIndexTask = new ActionListener<BulkByScrollResponse>() {
+                @Override
+                public void onResponse(BulkByScrollResponse bulkByScrollResponse) {
+                    try {
+                        log.info("Reindex from {} to {} succeeded", index, destIndex);
+                        renameReindexedIndices(searchClient, index, index + "*", destIndex);
+                        log.info("Finished setting up {}", index);
+                        MetricUtils.counter(this.getClass(), ES_REINDEX_SUCCESS_METRIC).inc();
+                    } catch (IOException e) {
+                        log.error(String.format("Failed to rename index %s after reindexing", index));
+                        MetricUtils.exceptionCounter(ESBulkProcessor.class, ES_REINDEX_FAILED_METRIC, e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    log.error(String.format("Failed to reindex %s", index));
+                    MetricUtils.exceptionCounter(ESBulkProcessor.class, ES_REINDEX_FAILED_METRIC, e);
+                }
+            };
+            Cancellable resp = searchClient.reindexAsync(reindexRequest, RequestOptions.DEFAULT, renameIndexTask);
```

Comment:
This is what I mentioned, I think that this is the less desirable async method as it doesn't give a task ID on the ES side. It just gives you back a Cancellable on the client side. There are a few problems with this:

1) The cancellability of the task lives on the client side in memory. If the application dies we now have a loose reindex that we can't control if I'm understanding what it's doing correctly. This isn't ideal as reindexing kind of hoses the index and we want to have external control to it.

2) It seems to indicate that this will be a blocking process on the ES side, I'm not sure what the implications of this are and if it could have more pronounced performance impacts on the rest of the indices being able to serve data. The docs are unclear here on what the difference is from what I've seen. Would like @david-leifker to weigh in on this, it's a gap in knowledge for me.

Reply:
I would like to change the order of the alias updates for this. First adjust the alias to the new index (temporarily having queries against it return no results), however it would receive any new writes immediately. Then the single async task is to backfill the older timeseries. If there is a way to order the reindex, even better since we can start with the most important recent docs. Maybe reindex by query + sort (not sure if it works). This approach immediately improves performance. Once under the threshold the overall process can then revert to the deleteByQuery method (after the most extreme case is solved with this method).